command/ca/policy/actions/dns.go

```diff
@@ -0,0 +1,182 @@
+'''
+$ step ca policy acme x509 allow dns "*.account1.acme.example.com" --provisioner my_acme_provisioner --reference account1
+'''
+
+Allow all DNS subdomains of "local" in SSH host certificates on authority level
+'''
+$ step ca policy authority ssh host allow dns "*.local"
+'''
+
+Deny badsshhost.local in SSH host certificates on authority level
+'''
+$ step ca policy authority ssh host allow dns "badsshhost.local"
+'''
+
+`, commandName),
+		Action: command.InjectContext(
+			ctx,
+			dnsAction,
+		),
+		Flags: []cli.Flag{
+			provisionerFilterFlag,
+			flags.EABKeyID,
+			flags.EABReference,
+			flags.AdminCert,
+			flags.AdminKey,
+			flags.AdminProvisioner,
+			flags.AdminSubject,
+			flags.PasswordFile,
+			flags.CaURL,
+			flags.Root,
+			flags.Context,
+			cli.BoolFlag{
+				Name:  "remove",
+				Usage: `removes the provided DNS names from the policy instead of adding them`,
+			},
+		},
+	}
+}
+
+func dnsAction(ctx context.Context) (err error) {
+
+	clictx := command.CLIContextFromContext(ctx)
+
+	args := clictx.Args()
+	if len(args) == 0 {
+		return errors.New("please provide at least one domain")
+	}
+
+	client, err := cautils.NewAdminClient(clictx)
+	if err != nil {
+		return fmt.Errorf("error creating admin client: %w", err)
+	}
+
+	policy, err := retrieveAndInitializePolicy(ctx, client)
+	if err != nil {
+		return fmt.Errorf("error retrieving policy: %w", err)
+	}
+
+	var dns []string
+
+	switch {
+	case policycontext.IsSSHHostPolicy(ctx):
+		switch {
+		case policycontext.IsAllow(ctx):
+			dns = policy.Ssh.Host.Allow.Dns
+		case policycontext.IsDeny(ctx):
+			dns = policy.Ssh.Host.Deny.Dns
+		default:
+			panic("no allow nor deny context set")
+		}
+	case policycontext.IsSSHUserPolicy(ctx):
+		return errors.New("SSH user policy does not support DNS names")
+	case policycontext.IsX509Policy(ctx):
+		switch {
+		case policycontext.IsAllow(ctx):
+			dns = policy.X509.Allow.Dns
+		case policycontext.IsDeny(ctx):
+			dns = policy.X509.Deny.Dns
+		default:
+			panic("no allow nor deny context set")
+		}
+	default:
+		panic("no SSH nor X.509 context set")
+	}
+
+	if clictx.Bool("remove") {
+		for _, domain := range args {
+			dns = remove(domain, dns)
+		}
+	} else {
+		dns = append(dns, args...)
+	}
+
+	switch {
+	case policycontext.IsSSHHostPolicy(ctx):
+		switch {
+		case policycontext.IsAllow(ctx):
+			policy.Ssh.Host.Allow.Dns = dns
+		case policycontext.IsDeny(ctx):
+			policy.Ssh.Host.Deny.Dns = dns
+		default:
+			panic("no allow nor deny context set")
+		}
+	case policycontext.IsSSHUserPolicy(ctx):
+		return errors.New("SSH user policy does not support DNS names")
+	case policycontext.IsX509Policy(ctx):
+		switch {
+		case policycontext.IsAllow(ctx):
+			policy.X509.Allow.Dns = dns
+		case policycontext.IsDeny(ctx):
+			policy.X509.Deny.Dns = dns
+		default:
+			panic("no allow nor deny context set")
+		}
+	default:
+		panic("no SSH nor X.509 context set")
+	}
```

Comment:
This switch and the above one can be converted to one moving the remove/append logic to a method and set these slices with the result of that method:
```golang
policy.xxx.Dns = addRemoveArguments(policy.xxx.Dns, args, ctx.Bool("remove"))
```

You can use the same command in other actions.

Reply:
Thanks. That cleaned it up 🙂